### PR TITLE
Send logs to Port integration

### PR DESCRIPTION
**Pull request by yairsimantov20**
# Description

What - Currently it is very hard to debug the ocean integration
Why - The integrations installed over a container which sometimes is not very convenient to get the logs from and everytime it restarts all the log history is lost
How - Added a log handler to capture logs from the integration and send them to Port

## Type of change
- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: integrations/jira/jira/client.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from httpx import Timeout, BasicAuth
 from jira.overrides import JiraResourceConfig
 from loguru import logger
+
```
Comment (yairsimantov20) [inline integrations/jira/jira/client.py:7]: revert me

File: port_ocean/consumers/kafka_consumer.py [hunk 2]
```diff
@@ -30,8 +29,8 @@ def __init__(
         self.org_id = org_id
         self.config = config
 
-        signal.signal(signal.SIGINT, self.exit_gracefully)
-        signal.signal(signal.SIGTERM, self.exit_gracefully)
+        # signal.signal(signal.SIGINT, self.exit_gracefully)
+        # signal.signal(signal.SIGTERM, self.exit_gracefully)
```
Comment (yairsimantov20) [inline port_ocean/consumers/kafka_consumer.py:34]: take care of this

File: port_ocean/ocean.py [hunk 3]
```diff
@@ -40,9 +42,9 @@ def __init__(
             base_path="./", **(config_override or {})
         )
         if config_factory:
-            self.config.integration.config = config_factory(
-                **self.config.integration.config
-            ).dict()
+            config, sensitive = config_factory(**self.config.integration.config)
+            self.config.integration.config = config.dict()
+            sensitive_log_filter.hide_sensitive_tokens(*sensitive)
```
Comment (yairsimantov20) [inline port_ocean/ocean.py:47]: hide ocean config sensitive configuration. CLIENT_ID SECRET etc...

File: port_ocean/ocean.py [hunk 4]
```diff
@@ -86,11 +87,12 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
 
         @self.fast_api_app.on_event("startup")
         async def startup() -> None:
+            set_signal_handler()
```
Comment (yairsimantov20) [inline port_ocean/ocean.py:90]: handle for kafka and webhook handlers

File: port_ocean/utils/signal.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+from loguru import logger
+from werkzeug.local import LocalProxy, LocalStack
+
+from port_ocean.utils.misc import generate_uuid
+
+
+class SignalHandler:
+    def __init__(self):
+        self._handlers: dict[str, Callable] = {}
```
Comment (Tankilevitch) [inline port_ocean/utils/signal.py:12]: Not especially related to the signal class, but I don't see that there is signal handling for all the event listeners ( kafka, once )

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Very nice job @yairsimantov20 🚀 🌊 
Left some comments

File: port_ocean/log/handlers.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from logging import LogRecord
+from logging.handlers import MemoryHandler
+
+from loguru import logger
+
+from port_ocean import Ocean
+from port_ocean.context.ocean import ocean
+from port_ocean.exceptions.context import PortOceanContextNotFoundError
+
+EMPTY_LOG_RECORD = LogRecord("", 0, "", 0, "", (), None)
```
[the remark this answers is not in the document]
Comment (yairsimantov20) [reply port_ocean/log/handlers.py:15]: not used anymore

File: port_ocean/log/handlers.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+
+class HTTPMemoryHandler(MemoryHandler):
+    def __init__(
+        self,
+        capacity: int,
+        flush_level: int = logging.FATAL,
+        flush_interval: int = 15,
+        flush_size: int = 2048,
+    ):
```
[the remark this answers is not in the document]
Comment (yairsimantov20) [reply port_ocean/log/handlers.py:25]: move the values to the init

File: port_ocean/utils/signal.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    return _signal_handler.top
+
+
+signal_handler = LocalProxy(_get_signal_handler)
+
+
+def init_signal_handler():
+    global _signal_handler
+    if _signal_handler.top is not None:
+        raise RuntimeError("Signal handler is already initialized")
```
Comment (yairsimantov20) [inline port_ocean/utils/signal.py:52]: change runtime error with custom ones

File: port_ocean/log/sensetive.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
+import re
+from typing import Callable, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from loguru import Record
+
+# https://github.com/h33tlit/secret-regex-list
+secret_patterns = {
+    "Cloudinary": r"cloudinary:\/\/.*",
```
Comment (Tankilevitch) [inline port_ocean/log/sensetive.py:9]: super cool, I am thinking maybe we would like to reduce the amount of patterns to those logically related to what ocean will support exporting. 

e.g. square, twitter sounds too far from integration that will be supported

File: port_ocean/config/base.py [hunk 10]
```diff
@@ -153,3 +156,29 @@ def customise_sources(  # type: ignore
...
                 ),
             )
+
+
+class BaseOceanModel(BaseModel):
+    def get_sensitive_fields_data(self) -> set[str]:
+        return _handle_get_sensitive_information(self)
+
+
+def _handle_get_sensitive_information(
```
Comment (Tankilevitch) [inline port_ocean/config/base.py:166]:
```suggestion
def _handle_sensitive_information(
```

File: port_ocean/ocean.py [hunk 11]
```diff
@@ -86,11 +94,12 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
 
         @self.fast_api_app.on_event("startup")
         async def startup() -> None:
+            init_signal_handler()
             try:
                 await self.integration.start()
                 await self._setup_scheduled_resync()
             except Exception as e:
-                logger.error(f"Failed to start integration with error: {e}")
+                logger.exception(f"Failed to start integration with error: {e}")
```
Comment (Tankilevitch) [inline port_ocean/ocean.py:102]: no need for e when using exception if IIRC

File: port_ocean/consumers/kafka_consumer.py [hunk 12]
```diff
@@ -1,12 +1,10 @@
-import signal
+import threading
 from typing import Any, Callable
 
 from confluent_kafka import Consumer, KafkaException, Message  # type: ignore
 from loguru import logger
 from pydantic import BaseModel
 
-from port_ocean.consumers.base_consumer import BaseConsumer
```
Comment (Tankilevitch) [inline port_ocean/consumers/kafka_consumer.py:8]: so `baseConsumer` can be deleted?

File: port_ocean/consumers/kafka_consumer.py [hunk 13]
```diff
@@ -64,8 +60,7 @@ def start(self) -> None:
                 ),
             )
             logger.info("Subscribed to topics")
-            self.running = True
-            while self.running:
+            while self.running and not event.is_set():
```
Comment (Tankilevitch) [inline port_ocean/consumers/kafka_consumer.py:63]: very cool

File: port_ocean/core/event_listener/base.py [hunk 14]
```diff
@@ -18,13 +22,29 @@ def __init__(
         events: EventListenerEvents,
     ):
         self.events = events
+        self._tasks_to_close: list[Task[Any]] = []
 
-    @abstractmethod
     async def start(self) -> None:
+        _id = signal_handler.register(self._stop)
```
Comment (Tankilevitch) [inline port_ocean/core/event_listener/base.py:28]: redundant `_id`?

File: port_ocean/core/event_listener/polling.py [hunk 15]
```diff
@@ -1,3 +1,5 @@
+import asyncio
+from asyncio import Task
```
Comment (Tankilevitch) [inline port_ocean/core/event_listener/polling.py:2]: choose one

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
minor

File: port_ocean/consumers/kafka_consumer.py (hunk 12 above)
Comment (yairsimantov20) [reply port_ocean/consumers/kafka_consumer.py:8]: seed
thought i already did

File: port_ocean/ocean.py [hunk 16]
```diff
@@ -39,10 +41,17 @@ def __init__(
...
-            ).dict()
+            raw_config = (
+                self.config.integration.config
+                if isinstance(self.config.integration.config, dict)
+                else self.config.integration.config.dict()
+            )
+            self.config.integration.config = config_factory(**raw_config)
+        sensitive_log_filter.hide_sensitive_tokens(
+            *self.config.get_sensitive_fields_data()
+        )
```
Comment (Tankilevitch) [inline port_ocean/ocean.py:54]:
```suggestion
        # add the integration sensitive configuration to the sensitive patterns to mask out 
        sensitive_log_filter.hide_sensitive_tokens(
            *self.config.get_sensitive_fields_data()
        )
```

**Review by Tankilevitch — APPROVED** [review]
LGTM 🌊

**Review by MatanHeledPort — APPROVED** [review]
lgtm

**Pull request merged**